JustSaying.IntegrationTests/Fluent/IntegrationTestBase.cs

```diff
@@ -80,7 +80,17 @@ protected async Task WhenAsync(IServiceCollection services, Func<IMessagePublish
 
             using (var source = new CancellationTokenSource(Timeout))
             {
-                await Task.WhenAny(action(publisher, listener, source.Token), Task.Delay(Timeout, source.Token)).ConfigureAwait(false);
+                var delayTask = Task.Delay(Timeout, source.Token);
+                var actionTask = action(publisher, listener, source.Token);
+
+                await Task.WhenAny(actionTask, delayTask).ConfigureAwait(false);
+
+                source.Token.ThrowIfCancellationRequested();
+
+                if (actionTask.IsFaulted)
+                {
+                    throw actionTask.Exception ?? new Exception("The action is faulted, but did not have an exception");
```

Comment:
Here it would be better to use `ExceptionDispatchInfo.Capture` & '.Throw()` so that the stacktrace is preserved. But it would be simpler to await the task to let the exception out.

Reply:
I think this covers all bases:
```csharp
using (var source = new CancellationTokenSource(Timeout))
{
    var actionTask = action(publisher, listener, source.Token);
    
    await Task.WhenAny(actionTask, Task.Delay(Timeout, source.Token))
                    .ConfigureAwait(false);
    
    source.Token.ThrowIfCancelled();
    source.Cancel();
    await actionTask.ConfigureAwait(false);
}
```

CTS is disposed, and cancelled to ensure there's no left over timers.
If it times out, `OperationCanceledException` is thrown. 
The `"The action is faulted, but did not have an exception"` case isn't a real concern as `Exception` will be non-null in the case `IsFaulted` is true.

Reply:
Interesting, I did not know about `ExceptionDispatchInfo.Capture` 

So what I'm seeing is that the stack trace from `WhenAsync` is on the `AggregateException`, and that wrapper contains the real exception with the correct stack trace. I'm happy with that.

Reply:
You're right about `new Exception(...)` being a nummer to make a warning go away. A faulted task has an exception.
So `source.Cancel()` breraks a few more tests, not doing that now.
I don't think `await actionTask` is safe in the general case (it might have hung) and you would be relying on having got past the timeout check at that point.